### ISSUE TITLE
Multiple Env In One Shot

### ISSUE THREAD

**Issue by AhmedNourJamalElDin** [Milad-Akarie/injectable#30]
Hi,
When I try to register one singleton for two different envs, it doesnt work. it seems I need to duplicate the code:
```dart
@registerModule
abstract class RegisterModule {
  @lazySingleton
  @prod
  @dev
  Service get service => ServiceImpl();
}
```
it only generates:
```dart
  //Register prod Dependencies --------
  if (environment == 'prod') {
    g.registerLazySingleton<Service>(
        () => registerModule.service);
  }
```
Expected behavior:
```dart
  //Register prod Dependencies --------
  if (environment == 'prod') {
    g.registerLazySingleton<Service>(
        () => registerModule.service);
  }
  if (environment == 'dev') {
    g.registerLazySingleton<Service>(
        () => registerModule.service);
  }
```

Am I missing something or is there a better way to do this?

Thanks :D

**Comment by Milad-Akarie** [comment]
Hey @AhmedNourJamalElDin Is there a good reason why do you need to generate the same exact singleton under different environments?

**Comment by AhmedNourJamalElDin** [comment]
Here is my case but IDK if it's the best or there is a better alternative (hope so):
I want to mock some `Service` for testing only, and I don't want to mock it for prod, staging, and dev.
when I use this code:
```dart
@registerModule
abstract class RegisterModule {
  // default for all stages except test
  @lazySingleton
  Service get service => ServiceImpl();

// only for test
  @lazySingleton
  @test
  Service get service => MockServiceImpl();
}
```
I got an error stating `Service` is already registered.
any better solution?

**Comment by guyluz11** [comment]
Can be the same as this ticket #17 .
@Milad-Akarie in my case I had class that needs too environments like dev and prod, and another class that use the test environment, and both of them are extending the same abstract class.
By changing the environment I expect it to point the the correct implementation of the abstract class.

**Comment by Milad-Akarie** [comment]
Hey @AhmedNourJamalElDin  this's how you actually supposed to do it at least for the current approach, but in your case if you set your environment to test it is going to be registered twice because deps without specific environments will always register.

I get what you're trying to do but it can't be done with what we have now..so you need to specify the env for each dependency.

**Comment by Milad-Akarie** [comment]
@guyluz11 
try this 

```dart
@injectable
@RegisterAs(Service, env: dev.name)
ServiceDevImpl extends Service {}

@injectable
@RegisterAs(Service, env: prod.name)
ServiceProdImpl extends Service {}

@injectable
@RegisterAs(Service, env: test.name)
ServiceTestImpl extends Service {}
```

or using the registerModule

```dart

@RegisterAs(Service, env: dev.name)
Service get ServiceDevImpl;

@RegisterAs(Service, env: prod.name)
Service get ServiceProdImpl;

@RegisterAs(Service, env: test.name)
Service get ServiceTestImpl;
```

**Comment by Milad-Akarie** [comment]
This

```dart
@RegisterAs(Service, env: dev.name)
```

is just a short for
```dart
@dev
@RegisterAs(Service,)
```

**Comment by guyluz11** [comment]
@Milad-Akarie This is a solution but duplication of code is not convenient. 
Will it be solved in upcoming versions?, I can endure it for a while but it is not good practice to duplicate the same code.

**Comment by Milad-Akarie** [comment]
There's no duplication here, you're registering different instances. the only way it can all be declared in one shot is by annotating the interface it self as it used to be in injectable 0.1.0 but that didn't work because an interface shouldn't know about it's implementations.

**Comment by guyluz11** [comment]
Is this example correct?

```
abstract class AnimalAbstract {
  int numberOfLegs();
}
```

```
@RegisterAs(AnimalAbstract, env: Env.prod)
@injectable
class CowAnimal extends AnimalAbstract {
  @override
  int numberOfLegs() {
    return 4;
  }

  @override
  String name() {
    return "cow";
  }
  
  void walk() {
    print("Walking");
  }
}


@RegisterAs(AnimalAbstract, env: Env.dev)
@injectable
class CowAnimal extends AnimalAbstract {
  @override
  int numberOfLegs() {
    return 4;
  }

  @override
  String name() {
    return "cow";
  }
  
  void walk() {
    print("Walking");
  }
}

@RegisterAs(AnimalAbstract, env: Env.test)
@injectable
class MockCowAnimal extends Mock implements AnimalAbstract {}
```

In this case `Cow` is not long but what if it is 500 lines? it mean that I have the same class twice with the same code which give us 1000 lines, and I you have four environments so now we have 2000 lines for the same code.
One change needs to be in four places.

**Comment by guyluz11** [comment]
What if we had something like  `@RegisterAs(AnimalAbstract, env: Env.test)` but `env` will get list of environments instead of one and it will create code like this 
`  if (environment == 'dev' || environment == 'prod' ...)`

**Comment by Milad-Akarie** [comment]
@guyluz11  I'm not sure if I understand currently but why would you need to register the same CowAnimal under different envs? if it's going to behave the same every-time.

**Comment by guyluz11** [comment]
@Milad-Akarie  Because I have MockCowAnimal which is register under test and is extending Mock from `mockito` package, this way when running the test environment all my app is using the mock implementation and I can test easily just the part that I need to test by calling it directly.

**Comment by AhmedNourJamalElDin** [comment]
> @Milad-Akarie Because I have MockCowAnimal which is register under test and is using Mock from `mockito` package as you can see in my example.
> This way when running the test environment all my app is using the mock implementation and I can test easily just the part that I need to test by calling it directly.

This is exactly what I'm looking for :D
Same service for dev, staging, and prod but mocked for test :D

**Comment by saschaernst** [comment]
> What if we had something like `@RegisterAs(AnimalAbstract, env: Env.test)` but `env` will get list of environments instead of one and it will create code like this
> ` if (environment == 'dev' || environment == 'prod' ...)`

To prevent breaking changes, what about introducing an annotation like 
```class Environments {
  final List<String> included;
  final List<String> excluded;

  const Environments({this.included, this.excluded});
}```

Could the generator handle that? (i.e. generating the conditions, checking contradictions)

**Comment by felixjunghans** [comment]
What I am currently doing to avoid having to write the same code two or three times is to write a DevImpl and a StageImpl and this extends ProdImpl.

So for @guyluz11 example: 

```
@RegisterAs(AnimalAbstract, env: Env.prod)
@injectable
class CowAnimal extends AnimalAbstract {
  @override
  int numberOfLegs() {
    return 4;
  }

  @override
  String name() {
    return "cow";
  }
  
  void walk() {
    print("Walking");
  }
}

@RegisterAs(AnimalAbstract, env: Env.dev)
@injectable
class DevCowAnimal extends CowAnimal {}

@RegisterAs(AnimalAbstract, env: Env.stage)
@injectable
class StageCowAnimal extends CowAnimal {}

@RegisterAs(AnimalAbstract, env: Env.test)
@injectable
class MockCowAnimal extends Mock implements AnimalAbstract {}
```

**Comment by felixjunghans** [comment]
@Milad-Akarie How about simply wrapping the register functions of getIt with a method that checks if the same type has already been registered.  

For example

```
  /// Only register if this type was not registered before.
  /// Prevents an error from GetIt that is thrown when a type is registered multiple times.
  void _registerIfAbsent<T>(Function register, {String instanceName}) {
    if (!g.isRegistered<T>()) {
      register();
    }
  }
```
A register function would look like this

```
  _registerIfAbsent<AnimalAbstract>(
      () => g.registerFactory<AnimalAbstract>(CowAnimal()));
```

Additionally you would only have to register the classes that have defined a special environment before the classes without environment at getIt.

Then it would be enough to define

```
@RegisterAs(AnimalAbstract)
@injectable
class CowAnimal extends AnimalAbstract {}
```

```
@RegisterAs(AnimalAbstract, env: Env.test)
@injectable
class MockCowAnimal extends Mock implements AnimalAbstract {}
```

The generated code would look like this

```
  //Register test Dependencies --------
  if (environment == 'test') {
    _registerIfAbsent<AnimalAbstract>(() => g.registerFactory<AnimalAbstract>(MockCowAnimal()));
  }

  _registerIfAbsent<AnimalAbstract>(() => g.registerFactory<AnimalAbstract>(CowAnimal()));
```

This would mean that the mock is instantiated in Test Environment and In all other cases CowAnimal would be instantiated. This would save a lot of unnecessary code and would also allow you to manually instantiate certain classes in GetIt before $initGetIt was called. For example in cases where you want to use the test environment but you want a certain class to behave differently for this test. This would make the injectable package even more flexible in my opinion.

**Comment by Milad-Akarie** [comment]
finally implemented in injectable: 1.0.0 check it out
